MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,960 @@
+`custom token`.
+
+### Permissions
+
+As mentioned, an account update may provide new values for certain account
+fields. Whether those changes are permitted is governed by a set of permissions.
+A permission may take on values `None`, `Either`, `Proof`, `Signature`, or
+`Impossible`, of OCaml type `Auth_required.t`.
+
+There is a set of associated tags `Proof`, `Signature`, and `None_given`. An
+account update that has a proof that verifies yields the tag `Proof`. A
+signature that verifies yields the tag `Signature`. If neither a proof nor a
+signature verifies, we get the `None_given` tag.
+
+Given a permission and a tag, an account update can perform an operation
+according to this OCaml function:
+
+```ocaml
+let check (t : Auth_required.t) (c : Control.Tag.t) =
+  match (t, c) with
+  | Impossible, _ ->
+      false
+  | None, _ ->
+      true
+  | Proof, Proof ->
+      true
+  | Signature, Signature ->
+      true
+  | Either, (Proof | Signature) ->
+      true
+  | Signature, Proof ->
+      false
+  | Proof, Signature ->
+      false
+  | (Proof | Signature | Either), None_given ->
+      false
+```
+
+The permissions that govern individual field updates are:
+
+- `edit_state`
+- `set_delegate`
+- `set_permissions`
+- `set_verification_key`
+- `set_zkapp_uri`
+- `edit_sequence_state`
+- `set_token_symbol`
+- `increment_nonce`
+- `set_voting_for`
+- `set_timing`
+
+There are permissions for sending and receiving tokens:
+
+- `send`
+- `receive`
+
+Finally, there is an `access` permission, which allows an account
+update to access a ledger account. That permission disallows all
+access to a ledger account if the authorization does not check against
+the permission, or the proof or signature authorization does not
+verify.
+
+The `access` permission introduced here affects not only zkApps, but
+also the payment and delegation transactions in the existing mainnet.
+For those transactions, the fee payer and source accounts (which are
+always the same accounts, in any case) need to have an `access`
+permission compatible with a signature authorization.
+
+### Processing zkApp transactions
+
+#### zkApp Transaction Entrypoints
+
+A user can create a zkApp transaction using SnarkyJS or other tooling, and sign
+it with a wallet key. The daemon provides a GraphQL endpoint to accept the
+zkApp. If the zkApp has a sufficient fee and passes validity checks (see below),
+it's queued to the transaction pool for broadcast to the network.
+
+#### Fees and queueing
+
+The fee payer's account update specifies a transaction fee and a
+nonce. The fee payer nonce must agree with the nonce in the fee
+payer's ledger account, or the zkApp is not added to the transaction
+pool. All signatures and proofs in the zkApp must verify before adding
+it to the pool. There is a size heuristic for zkApps: Those zkApps
+that are too large according to that heuristic are not accepted into
+the pool.
+
+The size heuristic involves three limits: a limit on the number of
```

Comment:
What is the proposal on the setting for the size heuristic?

Reply:
updated with specifics